In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime, date
%matplotlib inline

In [2]:
import seaborn as sns

# Customer Address Sheet

In [3]:
ca = pd.read_excel('Raw_data.xlsx', sheet_name = 'CustomerAddress')

In [4]:
ca.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [5]:
ca.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


The data type is correct. There are 3 numeric columns and 3 object type columns. We need to clean the data and check the data quality before using the data for further analysis. 

In [6]:
ca.shape

(3999, 6)

### Checking Missing Values

In [7]:
#missing value check
ca.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

#### No missing values found

### Performing Duplication Check 

In [8]:
ca = ca.drop_duplicates()

In [9]:
ca

customer_id                    address  postcode            state  \
0               1         060 Morning Avenue      2016  New South Wales   
1               2        6 Meadow Vale Court      2153  New South Wales   
2               4         0 Holy Cross Court      4211              QLD   
3               5        17979 Del Mar Point      2448  New South Wales   
4               6           9 Oakridge Court      3216              VIC   
...           ...                        ...       ...              ...   
3994         3999            1482 Hauk Trail      3064              VIC   
3995         4000  57042 Village Green Point      4511              QLD   
3996         4001     87 Crescent Oaks Alley      2756              NSW   
3997         4002           8194 Lien Street      4032              QLD   
3998         4003            320 Acker Drive      2251              NSW   

        country  property_valuation  
0     Australia                  10  
1     Australia                  10  
2     Australia                   9  
3     Australia                   4  
4     Australia                   9  
...         ...                 ...  
3994  Australia                   3  
3995  Australia                   6  
3996  Australia                  10  
3997  Australia                   7  
3998  Australia                   7  

[3999 rows x 6 columns]

#### Since the total no. of rows are same in both cases then the table has no duplicates

### Performing Consistency Check

In [10]:
ca['state'].value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

New South Wales and Victoria occur twice and in different forms

In [11]:
ca['state'].replace('New South Wales','NSW',inplace = True)
ca['state'].replace('Victoria','VIC',inplace = True)

In [12]:
ca['state'].value_counts()

NSW    2140
VIC    1021
QLD     838
Name: state, dtype: int64

In [13]:
ca['country'].value_counts()

Australia    3999
Name: country, dtype: int64

In [14]:
ca[['address', 'postcode']].drop_duplicates()

address  postcode
0            060 Morning Avenue      2016
1           6 Meadow Vale Court      2153
2            0 Holy Cross Court      4211
3           17979 Del Mar Point      2448
4              9 Oakridge Court      3216
...                         ...       ...
3994            1482 Hauk Trail      3064
3995  57042 Village Green Point      4511
3996     87 Crescent Oaks Alley      2756
3997           8194 Lien Street      4032
3998            320 Acker Drive      2251

[3999 rows x 2 columns]

No inconsistency in postcode column. I have completed data cleaning on customer address sheet.

In [15]:
ca.to_csv('Clean_CustomerAddress.csv', index = False)

# Customer Demographic Sheet

In [18]:
cd = pd.read_excel('Raw_data.xlsx', sheet_name = 'CustomerDemographic')

In [17]:
import warnings
warnings.filterwarnings('ignore')

In [19]:
cd.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [20]:
cd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

The default column has random values and is of no use to us. So, we will drop it. 

In [22]:
#numeric columns
cd_num = cd.select_dtypes(include = [np.number])
cd_num.columns.values

array(['customer_id', 'past_3_years_bike_related_purchases', 'tenure'],
      dtype=object)

In [23]:
#Dropping default column
cd.drop('default',axis = 1, inplace = True)

### Checking Missing Values

In [24]:
cd.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                  87
dtype: int64

In [25]:
cd.isnull().mean()*100

customer_id                             0.000
first_name                              0.000
last_name                               3.125
gender                                  0.000
past_3_years_bike_related_purchases     0.000
DOB                                     2.175
job_title                              12.650
job_industry_category                  16.400
wealth_segment                          0.000
deceased_indicator                      0.000
owns_car                                0.000
tenure                                  2.175
dtype: float64

Checking if the customer id and first name are null when last name is null

In [26]:
nul_last_cd = cd[cd['last_name'].isnull()]
nul_last_cd[['customer_id','first_name']].isnull().sum()

customer_id    0
first_name     0
dtype: int64

This implies that all customers are identifiable. So, filling the value with "None".

In [27]:
cd['last_name'].fillna("None", axis = 0, inplace = True)

In [28]:
cd['last_name'].isnull().sum()

0

All missing values from the last name column has been dealt with.

In [33]:
cd[cd['DOB'].isnull()]

customer_id first_name   last_name gender  \
143           144       Jory  Barrabeale      U   
167           168     Reggie   Broggetti      U   
266           267      Edgar     Buckler      U   
289           290    Giorgio      Kevane      U   
450           451     Marlow   Flowerdew      U   
...           ...        ...         ...    ...   
3778         3779      Ulick      Daspar      U   
3882         3883      Nissa      Conrad      U   
3930         3931      Kylie       Epine      U   
3934         3935     Teodor   Alfonsini      U   
3997         3998     Sarene     Woolley      U   

      past_3_years_bike_related_purchases DOB                   job_title  \
143                                    71 NaT          Environmental Tech   
167                                     8 NaT             General Manager   
266                                    53 NaT                         NaN   
289                                    42 NaT      Senior Sales Associate   
450                                    37 NaT  Quality Control Specialist   
...                                   ...  ..                         ...   
3778                                   68 NaT                         NaN   
3882                                   35 NaT             Legal Assistant   
3930                                   19 NaT                         NaN   
3934                                   72 NaT                         NaN   
3997                                   60 NaT           Assistant Manager   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
143                     IT      Mass Customer                  N       No   
167                     IT  Affluent Customer                  N      Yes   
266                     IT     High Net Worth                  N       No   
289                     IT      Mass Customer                  N       No   
450                     IT     High Net Worth                  N       No   
...                    ...                ...                ...      ...   
3778                    IT  Affluent Customer                  N       No   
3882                    IT      Mass Customer                  N       No   
3930                    IT     High Net Worth                  N      Yes   
3934                    IT     High Net Worth                  N      Yes   
3997                    IT     High Net Worth                  N       No   

      tenure  
143      NaN  
167      NaN  
266      NaN  
289      NaN  
450      NaN  
...      ...  
3778     NaN  
3882     NaN  
3930     NaN  
3934     NaN  
3997     NaN  

[87 rows x 12 columns]

Earlier, the percentage of null values in DOB came out to be 2.175%. Since this is less than 5%, we remove the records where dob is null 

In [34]:
dob_null = cd[cd['DOB'].isnull()].index

cd.drop(dob_null, axis = 0, inplace = True)

In [35]:
cd['DOB'].isnull().sum()

0

Missing values from the DOB column has now become 0

#### Checking for further descripency by calculating age

In [42]:
def Age(birthdate):
    today = date.today() - timedelta(days = 7*365)
    return today.year - birthdate.year - ((today.month, today.day) < (birthdate.month, birthdate.day))

cd['age'] = cd['DOB'].apply(Age)

In [37]:
from datetime import timedelta

In [43]:
cd['age'].describe()

count    3913.000000
mean       39.948377
std        12.796020
min        15.000000
25%        30.000000
50%        40.000000
75%        49.000000
max       174.000000
Name: age, dtype: float64

In [44]:
#checking for outliers
cd[cd['age']>100]

customer_id first_name last_name gender  \
33           34   Jephthah  Bachmann      U   

    past_3_years_bike_related_purchases        DOB        job_title  \
33                                   59 1843-12-21  Legal Assistant   

   job_industry_category     wealth_segment deceased_indicator owns_car  \
33                    IT  Affluent Customer                  N       No   

    tenure  age  
33    20.0  174

Since there is a record where the age is 174 which is not possible so dropping it.

In [45]:
cd.drop(cd[cd['age']>100].index, axis = 0, inplace = True)

In [46]:
cd[cd['age']>100]

Empty DataFrame
Columns: [customer_id, first_name, last_name, gender, past_3_years_bike_related_purchases, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, owns_car, tenure, age]
Index: []

The DOB column has been cleaned.

In [47]:
cd['job_title'].isnull().mean()*100

12.704498977505112

Since the percentage is quite high, we replace the values with 'Not Provided'

In [48]:
cd['job_title'].fillna('Not Provided', axis = 0, inplace = True)

In [49]:
cd['job_title'].isnull().sum()

0

In [50]:
cd['job_industry_category'].isnull().mean()*100

16.768916155419223

In [52]:
cd['job_industry_category'].fillna('Not Provided', axis = 0,inplace = True)

In [54]:
cd['job_industry_category'].isnull().sum()

0

In [56]:
# Checking tenure for null values
cd['tenure'].isnull().sum()

0

### Performing Quality Check

In [57]:
# non-numeric columns
cd_non_num = cd.select_dtypes(exclude = np.number)
cd_non_num.columns.values

array(['first_name', 'last_name', 'gender', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'owns_car'], dtype=object)

Already checked first name and last name in the beginning. 

In [58]:
cd['gender'].value_counts()

Female    2037
Male      1872
F            1
Femal        1
M            1
Name: gender, dtype: int64

In [59]:
def rep_gender_name(gender):
    if gender == 'F':
        return 'Female'
    elif gender == 'Femal':
        return 'Female'
    elif gender == 'M':
        return 'Male'
    else:
        return gender
    
cd['gender'] = cd['gender'].apply(rep_gender_name)

In [60]:
cd['gender'].value_counts()

Female    2039
Male      1873
Name: gender, dtype: int64

In [61]:
cd['job_title'].value_counts()

Not Provided                            497
Business Systems Development Analyst     43
Tax Accountant                           43
Social Worker                            42
Recruiting Manager                       41
                                       ... 
Database Administrator II                 4
Health Coach III                          3
Health Coach I                            3
Research Assistant III                    3
Developer I                               1
Name: job_title, Length: 196, dtype: int64

In [62]:
cd['job_industry_category'].value_counts()

Manufacturing         796
Financial Services    767
Not Provided          656
Health                596
Retail                358
Property              267
IT                    151
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

In [63]:
cd['wealth_segment'].value_counts()

Mass Customer        1954
High Net Worth        996
Affluent Customer     962
Name: wealth_segment, dtype: int64

In [64]:
cd['deceased_indicator'].value_counts()

N    3910
Y       2
Name: deceased_indicator, dtype: int64

In [65]:
cd['owns_car'].value_counts()

Yes    1974
No     1938
Name: owns_car, dtype: int64

There is inconsistent data in job_title, wealth_segment,deceased_indicator and owns_car columns. 

In [66]:
cd = cd.drop_duplicates()

In [67]:
cd.shape

(3912, 13)

In [68]:
cd.to_csv('Clean_CustomerDemographic.csv',index = False)

# Transactions Sheet

In [3]:
tran = pd.read_excel('Raw_data.xlsx', sheet_name = 'Transactions')

In [70]:
tran.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [71]:
tran.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [72]:
tran.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

In [4]:
#Converting product_first_sold_date to datetime
tran['product_first_sold_date'] = pd.to_datetime(tran['product_first_sold_date']).dt.date

In [77]:
tran.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost product_first_sold_date  
0       71.49          53.62              1970-01-01  
1     2091.47         388.92              1970-01-01  
2     1793.43         248.82              1970-01-01  
3     1198.46         381.10              1970-01-01  
4     1765.30         709.48              1970-01-01

In [78]:
tran.shape

(20000, 13)

### Checking Missing Values

In [5]:
tran.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [80]:
tran.isnull().mean()*100

transaction_id             0.000
product_id                 0.000
customer_id                0.000
transaction_date           0.000
online_order               1.800
order_status               0.000
brand                      0.985
product_line               0.985
product_class              0.985
product_size               0.985
list_price                 0.000
standard_cost              0.985
product_first_sold_date    0.985
dtype: float64

In [81]:
tran[tran['online_order'].isnull()]

transaction_id  product_id  customer_id transaction_date  online_order  \
97                 98          49          333       2017-06-23           NaN   
166               167          90         3177       2017-04-26           NaN   
169               170           6          404       2017-10-16           NaN   
250               251          63         1967       2017-04-11           NaN   
300               301          78         2530       2017-03-24           NaN   
...               ...         ...          ...              ...           ...   
19514           19515          51          690       2017-01-22           NaN   
19573           19574          18         1735       2017-01-15           NaN   
19580           19581          49         1933       2017-10-12           NaN   
19635           19636          98         1389       2017-07-26           NaN   
19843           19844           3            8       2017-01-28           NaN   

      order_status           brand product_line product_class product_size  \
97        Approved   Trek Bicycles         Road        medium       medium   
166       Approved  Norco Bicycles     Standard           low       medium   
169       Approved      OHM Cycles     Standard          high       medium   
250       Approved           Solex     Standard        medium       medium   
300       Approved  Giant Bicycles     Standard        medium        large   
...            ...             ...          ...           ...          ...   
19514     Approved      OHM Cycles     Standard          high       medium   
19573     Approved           Solex     Standard        medium       medium   
19580     Approved   Trek Bicycles         Road        medium       medium   
19635     Approved   Trek Bicycles     Standard          high       medium   
19843     Approved   Trek Bicycles     Standard        medium        large   

       list_price  standard_cost product_first_sold_date  
97         533.51         400.13              1970-01-01  
166        363.01         290.41              1970-01-01  
169        227.88         136.73              1970-01-01  
250       1483.20          99.59              1970-01-01  
300       1765.30         709.48              1970-01-01  
...           ...            ...                     ...  
19514     2005.66        1203.40              1970-01-01  
19573      575.27         431.45              1970-01-01  
19580      533.51         400.13              1970-01-01  
19635      358.39         215.03              1970-01-01  
19843     2091.47         388.92              1970-01-01  

[360 rows x 13 columns]

In [6]:
mode_online_order = tran['online_order'].mode()

In [7]:
mode_online_order

0    1.0
dtype: float64

Since we have price value associated online_order null values that means the customer did buy something.
So, we do mode imputation here.

In [8]:
tran['online_order'].replace(np.nan,1,inplace = True)

In [9]:
tran['online_order'].isnull().sum()

0

In [10]:
tran[tran['brand'].isnull()]

transaction_id  product_id  customer_id transaction_date  online_order  \
136               137           0          431       2017-09-23           0.0   
159               160           0         3300       2017-08-27           0.0   
366               367           0         1614       2017-03-10           0.0   
406               407           0         2559       2017-06-14           1.0   
676               677           0         2609       2017-07-02           0.0   
...               ...         ...          ...              ...           ...   
19340           19341           0          443       2017-12-26           1.0   
19383           19384           0         2407       2017-06-11           0.0   
19793           19794           0         2860       2017-01-13           0.0   
19859           19860           0         2468       2017-06-24           1.0   
19871           19872           0           61       2017-03-17           1.0   

      order_status brand product_line product_class product_size  list_price  \
136       Approved   NaN          NaN           NaN          NaN     1942.61   
159       Approved   NaN          NaN           NaN          NaN     1656.86   
366       Approved   NaN          NaN           NaN          NaN      850.89   
406       Approved   NaN          NaN           NaN          NaN      710.59   
676       Approved   NaN          NaN           NaN          NaN     1972.01   
...            ...   ...          ...           ...          ...         ...   
19340     Approved   NaN          NaN           NaN          NaN      744.54   
19383     Approved   NaN          NaN           NaN          NaN     1098.18   
19793     Approved   NaN          NaN           NaN          NaN      868.56   
19859     Approved   NaN          NaN           NaN          NaN     1497.43   
19871     Approved   NaN          NaN           NaN          NaN      867.92   

       standard_cost product_first_sold_date  
136              NaN                     NaT  
159              NaN                     NaT  
366              NaN                     NaT  
406              NaN                     NaT  
676              NaN                     NaT  
...              ...                     ...  
19340            NaN                     NaT  
19383            NaN                     NaT  
19793            NaN                     NaT  
19859            NaN                     NaT  
19871            NaN                     NaT  

[197 rows x 13 columns]

In [11]:
tran[tran['brand'].isnull()].isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order                 0
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

From the above two cells, we gather that when brand is null then product_line, product_class, product_size, standard_cost, product_first_sold_date columns are also null for the corresponding null values in the brand column.             

Therefore, we drop the rows where brand is null since the null% is 0.985% which is quite small. 

In [12]:
tran.drop(tran[tran['brand'].isnull()].index, axis = 0, inplace = True)

In [13]:
tran.isnull().sum()

transaction_id             0
product_id                 0
customer_id                0
transaction_date           0
online_order               0
order_status               0
brand                      0
product_line               0
product_class              0
product_size               0
list_price                 0
standard_cost              0
product_first_sold_date    0
dtype: int64

In [14]:
tran.shape

(19803, 13)

### Performing Quality checks

In [15]:
#non-numeric columns
tran_non_num = tran.select_dtypes(exclude = np.number)
tran_non_num.columns.values

array(['transaction_date', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'product_first_sold_date'],
      dtype=object)

In [16]:
tran['online_order'].value_counts()

1.0    10097
0.0     9706
Name: online_order, dtype: int64

In [17]:
tran['order_status'].value_counts()

Approved     19625
Cancelled      178
Name: order_status, dtype: int64

In [18]:
tran[['order_status', 'online_order']].drop_duplicates()

order_status  online_order
0       Approved           0.0
1       Approved           1.0
42     Cancelled           0.0
254    Cancelled           1.0

In [19]:
tran['brand'].value_counts()

Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: brand, dtype: int64

In [20]:
tran['product_line'].value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

In [21]:
tran['product_class'].value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [22]:
tran['product_size'].value_counts()

medium    12990
large      3976
small      2837
Name: product_size, dtype: int64

There is no problem with the data in order_status, online_order, brand, product_line, product_class, and product_size columns.

In [23]:
tran_after_removing_duplicate = tran.drop('transaction_id', axis=1).drop_duplicates()


In [24]:
tran_after_removing_duplicate.shape

(19803, 12)

In [25]:
tran.shape

(19803, 13)

Last two cells tells us that there are no duplicate rows in data after removing the primary key

In [26]:
tran.to_csv('Clean_Transactions.csv', index = False)

In [2]:
import math
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime, date